Proposal for docker run retry.

xquek opened the pull request:
@orodeh I was wondering if this is a good way to let docker pull retry a few times. Also, can I check how will you recommend I should test this?

I'm thinking 

1) Create a wdl task with a docker image that does not exist. Run the task and let it fail - check the log that it did retry 5 times

or 

2) create a wdl task with a docker image that does not exists. create docker submit script. run docker submit script and let it fail and check log that it retried 5 times

orodeh commented:
In terms of testing, method (1) is good. I think there is a test like this already (test/draft2/bad_status2.wdl).

In terms of the code, you identified the embedded shell script that needs to change. That's good work. The thing is, that with this change, the entire user bash script will be retried, not just the docker pull part. I would suggest splitting apart the two elements. First, do a docker pull that can handle failures. Then, run the script.

xquek commented:
Hi Ohad, thanks for the feedback. 

Currently, it only retries on the docker pull block. 
I think it is a good idea to keep them separate as well - will work on it next.
I will take a look at the test as well.

orodeh commented:
I believe the code, as it stands, will do the `docker run` block 5 times, even if the command is successful. I am suggesting splitting it into `docker pull`, and then `docker run`. The `docker pull` can be retried, in case of failure, up to five times.  

We can talk in person, if this isn't making sense.

xquek commented:
@orodeh I changed the code to pull and retry earlier. 
Added a test to pull an image that exists and an image that doesn't.
It will catch the exception and check the error message.  

Let me know if it makes sense. Thanks!

orodeh commented on src/main/scala/dxWDL/exec/TaskRunner.scala:
```diff
@@ -171,6 +171,30 @@ case class TaskRunner(task: CallableTaskDefinition,
         }
     }
 
+    private def pullImage(dImg: String, retry_count: Int): Option[String] = {
```
This is currently written as a recursive method. I suggest refactoring as a for loop. The recursive call is made inside the exception catch clause, which is more complex than you actually need.

orodeh commented on src/main/scala/dxWDL/exec/TaskRunner.scala:
```diff
@@ -182,17 +206,17 @@ case class TaskRunner(task: CallableTaskDefinition,
                 Utils.appletLog(verbose, s"looking up dx:url ${url}")
                 val dxFile = DxPath.resolveDxURLFile(url)
-		val fileName = dxFile.describe().getName
+		            val fileName = dxFile.describe().getName
```
This spaces are strange. Are you tabs converted into white space?

orodeh commented on src/main/scala/dxWDL/exec/TaskRunner.scala:
```diff
@@ -206,7 +230,11 @@ case class TaskRunner(task: CallableTaskDefinition,
                 Some(repo)
 
-            case _ => dImg
+            case Some(dImg) =>
```
This is elegant.

orodeh commented on src/test/scala/dxWDL/exec/TaskRunnerTest.scala:
```diff
@@ -250,4 +250,18 @@ class TaskRunnerTest extends FlatSpec with Matchers {
         runTask("missing_optional_output_file")
     }
+
+    it should "retry docker at least 5 times" taggedAs(EdgeTest) in {
```
This test is better done as a standalone WDL task. The general rule is to have short tests in the unit-test section, but this could take a significant amount of time.

xquek commented:
Removed tab character.
Converted from recursion to while loop.

job that failed -[ job-Ff71xq80jy8gpbByK6y09G8P](https://platform.dnanexus.com/projects/FJ90qPj0jy8zYvVV9yz3F5gv/monitor/job/Ff71xq80jy8gpbByK6y09G8P) - logs show that docker pull was run 5 times.

docker still able to pull [/job-Ff71xqQ0jy8g87gy5jqZgFk7](https://platform.dnanexus.com/projects/FJ90qPj0jy8zYvVV9yz3F5gv/monitor/job/Ff71xqQ0jy8g87gy5jqZgFk7)
